core/embed/rust/src/ui/display/image.rs

```diff
@@ -0,0 +1,220 @@
+        let height = image.read_u16_le(6)?;
+        let len = image.read_u32_le(8)? as usize;
+
+        if width > 1024 || height > 1024 || len > 65536 {
+            return None;
+        }
+
+        if len + Self::HEADER_LENGTH != image.len() {
+            return None;
+        }
+
+        Some(Self {
+            format,
+            size: Offset::new(width as i16, height as i16),
+            len,
+        })
+    }
+
+    pub fn format(&self) -> ToifFormat {
+        self.format
+    }
+
+    pub fn size(&self) -> Offset {
+        self.size
+    }
+
+    pub fn width(&self) -> i16 {
+        self.size.x
+    }
+
+    pub fn height(&self) -> i16 {
+        self.size.y
+    }
+
+    pub fn is_grayscale(&self) -> bool {
+        matches!(
+            self.format,
+            ToifFormat::GrayScaleOH | ToifFormat::GrayScaleEH
+        )
+    }
+
+    pub fn stride(&self) -> usize {
+        if self.is_grayscale() {
+            (self.width() + 1) as usize / 2
+        } else {
+            self.width() as usize * 2
+        }
+    }
+}
+
+pub struct JpegInfo {
+    size: Offset,
+    mcu_height: i16,
+}
+
+impl JpegInfo {
+    pub fn parse(image: BinaryData) -> Option<Self> {
+        const M_SOI: u16 = 0xFFD8;
+        const M_SOF0: u16 = 0xFFC0;
+        const M_DRI: u16 = 0xFFDD;
+        const M_RST0: u16 = 0xFFD0;
+        const M_RST7: u16 = 0xFFD7;
+        const M_SOS: u16 = 0xFFDA;
+        const M_EOI: u16 = 0xFFD9;
+
+        let mut result = None;
+        let mut ofs = 0;
+
+        while image.read_u16_be(ofs)? != M_SOI {
+            ofs += 1;
+        }
+
+        loop {
+            let marker = image.read_u16_be(ofs)?;
+
+            if (marker & 0xFF00) != 0xFF00 {
+                return None;
+            }
+
+            ofs += 2;
+
+            ofs += match marker {
+                M_SOI => 0,
+                M_SOF0 => {
+                    let w = image.read_u16_be(ofs + 3)? as i16;
+                    let h = image.read_u16_be(ofs + 5)? as i16;
+                    // Number of components
+                    let nc = image.read_u8(ofs + 7)?;
+                    if (nc != 1) && (nc != 3) {
+                        return None;
+                    }
+                    // Sampling factor of the first component
+                    let c1 = image.read_u8(ofs + 9)?;
+                    if (c1 != 0x11) && (c1 != 0x21) & (c1 != 0x22) {
+                        return None;
+                    };
+                    let mcu_height = (8 * (c1 & 15)) as i16;
+                    result = Some(JpegInfo {
+                        size: Offset::new(w, h),
+                        mcu_height,
+                    });
+
+                    image.read_u16_be(ofs)?
+                }
+                M_DRI => 4,
+                M_EOI => return None,
+                M_RST0..=M_RST7 => 0,
+                M_SOS => break,
+                _ => image.read_u16_be(ofs)?,
+            } as usize;
+        }
```

Comment:
Yes, I agree that we should develop a more universal reader and consider modifying `InputStream` to work with `BinaryData` objects. However, instead of implementing another reader right now, I would prefer to do it correctly and later. It seems challenging since `InputStream` is tailored to work only with slices. Maybe we could create an issue for this?